File: apps/web/src/store/auth-store.ts
import { create } from 'zustand';
import type { User, AuthChangeEvent, Session } from '@supabase/supabase-js';
import type { Profile } from '@raichu/shared-types';
import { getSupabaseBrowserClient } from '../lib/supabase/client';
import { analytics } from '../lib/analytics';

interface AuthStore {
  user: User | null;
  profile: Profile | null;
  loading: boolean;
  initialized: boolean;

  initialize: () => Promise<void>;
  signUp: (email: string, password: string, username: string) => Promise<void>;
  signInWithPassword: (email: string, password: string) => Promise<void>;
  signInWithGoogle: () => Promise<void>;
  signOut: () => Promise<void>;
  updateProfile: (data: Partial<Pick<Profile, 'display_name' | 'avatar_url'>>) => Promise<void>;
  fetchProfile: () => Promise<void>;
}

export const useAuthStore = create<AuthStore>((set, get) => ({
  user: null,
  profile: null,
  loading: false,
  initialized: false,

  initialize: async () => {
    const supabase = getSupabaseBrowserClient();

    // Get current session
    const {
      data: { session },
    } = await supabase.auth.getSession();

    if (session?.user) {
      set({ user: session.user });
      await get().fetchProfile();
    }

    // Listen for auth changes
    supabase.auth.onAuthStateChange(async (_event: AuthChangeEvent, session: Session | null) => {
      set({ user: session?.user ?? null });

      if (session?.user) {
        await get().fetchProfile();
      } else {
        set({ profile: null });
      }
    });

    set({ initialized: true });
  },

  signUp: async (email, password, username) => {
    set({ loading: true });
    try {
      const supabase = getSupabaseBrowserClient();
      const { data, error } = await supabase.auth.signUp({
        email,
        password,
        options: {
          data: { username },
        },
      });
      if (error) throw error;

      // If auto-confirm is on, a session is returned immediately
      if (data.session?.user) {
        set({ user: data.session.user });
        await get().fetchProfile();
        analytics.signup(data.session.user.id);
      }
    } finally {
      set({ loading: false });
    }
  },

  signInWithPassword: async (email, password) => {
    set({ loading: true });
    try {
      const supabase = getSupabaseBrowserClient();
      const { data, error } = await supabase.auth.signInWithPassword({ email, password });
      if (error) throw error;

      if (data.user) {
        set({ user: data.user });
        await get().fetchProfile();
        analytics.login(data.user.id);
      }
    } finally {
      set({ loading: false });
    }
  },

  signInWithGoogle: async () => {
    set({ loading: true });
    try {
      const supabase = getSupabaseBrowserClient();
      const { error } = await supabase.auth.signInWithOAuth({
        provider: 'google',
        options: {
          redirectTo: `${window.location.origin}/auth/callback`,
          queryParams: {
            access_type: 'offline',
            prompt: 'consent',
          },
        },
      });
      if (error) throw error;
    } finally {
      set({ loading: false });
    }
  },

  signOut: async () => {
    const { user } = get();
    analytics.logout(user?.id ?? null);
    const supabase = getSupabaseBrowserClient();
    await supabase.auth.signOut();
    set({ user: null, profile: null });
  },

  updateProfile: async (data) => {
    const { user } = get();
    if (!user) return;

    const supabase = getSupabaseBrowserClient();
    const { error } = await supabase
      .from('profiles')
      .update(data)
      .eq('id', user.id);

    if (error) throw error;
    await get().fetchProfile();
  },

  fetchProfile: async () => {
    const { user } = get();
    if (!user) return;

    const supabase = getSupabaseBrowserClient();
    const { data, error } = await supabase
      .from('profiles')
      .select('*')
      .eq('id', user.id)
      .single();

    if (!error && data) {
      set({ profile: data as Profile });
    }
  },
}));

File: apps/Ios-app/Raichu/Raichu/Stores/AuthStore.swift
// AuthStore.swift
// Raichu
// Authentication state: session, profile, sign in/up/out.

import SwiftUI
import Combine
import Supabase

@MainActor
final class AuthStore: ObservableObject {
    @Published var session: Session? = nil
    @Published var profile: Profile? = nil
    @Published var initialized: Bool = false
    @Published var loading: Bool = false
    @Published var error: String? = nil

    var isAuthenticated: Bool { session != nil }
    var userId: String? { session?.user.id.uuidString }
    var accessToken: String? { session?.accessToken }

    func initialize() async {
        // Restore existing session from Keychain (supabase-swift handles storage)
        session = try? await supabase.auth.session

        // Load cached profile immediately for offline display (spec 6.3)
        profile = loadCachedProfile()

        if session != nil {
            await fetchProfile()
        }

        // Listen for future auth state changes
        Task {
            for await (_, newSession) in supabase.auth.authStateChanges {
                self.session = newSession
                if newSession != nil {
                    await self.fetchProfile()
                } else {
                    self.profile = nil
                }
            }
        }

        initialized = true
    }

    func signIn(email: String, password: String) async throws {
        loading = true
        error = nil
        defer { loading = false }

        do {
            let result = try await supabase.auth.signIn(email: email, password: password)
            session = result
            await fetchProfile()
            // Request push notification permission on first sign-in (spec 6.4 item 1)
            NotificationManager.shared.requestPermission()
        } catch {
            self.error = Self.message(for: error)
            throw error
        }
    }

    func signUp(email: String, password: String, username: String) async throws {
        loading = true
        error = nil
        defer { loading = false }

        do {
            let result = try await supabase.auth.signUp(
                email: email,
                password: password,
                data: ["username": .string(username)]
            )
            session = result.session
            if result.session != nil {
                await fetchProfile()
                // Request push notification permission on first sign-up (spec 6.4 item 1)
                NotificationManager.shared.requestPermission()
            }
        } catch {
            self.error = Self.message(for: error)
            throw error
        }
    }

    func signInWithGoogle() async throws {
        error = nil
        do {
            // Use in-app browser so the user stays in the app during OAuth.
            try await supabase.auth.signInWithOAuth(
                provider: .google,
                redirectTo: AppConfig.oauthCallbackURL
            )
        } catch {
            self.error = Self.message(for: error)
            throw error
        }
    }

    /// Completes an OAuth sign-in from the redirect back into the app. Called
    /// from `RaichuApp`'s `onOpenURL`; without it the callback is dropped and
    /// the user never gets signed in.
    func handleOpenURL(_ url: URL) async {
        guard url.scheme == AppConfig.urlScheme else { return }
        do {
            session = try await supabase.auth.session(from: url)
            await fetchProfile()
        } catch {
            self.error = Self.message(for: error)
        }
    }

    /// Supabase surfaces most failures as opaque decoding or API errors; this
    /// keeps the user-facing copy readable without swallowing the detail.
    private static func message(for error: Error) -> String {
        let described = (error as NSError).localizedDescription
        return described.isEmpty ? "Something went wrong. Please try again." : described
    }

    func signOut() async {
        try? await supabase.auth.signOut()
        session = nil
        profile = nil
        // Clear cached profile on sign out
        UserDefaults.standard.removeObject(forKey: "cachedProfile")
    }

    func fetchProfile() async {
        guard let userId else { return }

        do {
            let p: Profile = try await supabase
                .from("profiles")
                .select()
                .eq("id", value: userId)
                .single()
                .execute()
                .value
            profile = p
            // Cache profile for offline display (spec 6.3)
            if let data = try? JSONEncoder().encode(p) {
                UserDefaults.standard.set(data, forKey: "cachedProfile")
            }
        } catch {
            // Profile fetch failure is non-fatal — use cached value if present
        }
    }

    // MARK: - Cache helpers

    private func loadCachedProfile() -> Profile? {
        guard let data = UserDefaults.standard.data(forKey: "cachedProfile") else { return nil }
        return try? JSONDecoder().decode(Profile.self, from: data)
    }

    func updateProfile(displayName: String? = nil, avatarUrl: String? = nil) async throws {
        guard let userId else { return }

        var updates: [String: AnyJSON] = [:]
        if let displayName { updates["display_name"] = .string(displayName) }
        if let avatarUrl { updates["avatar_url"] = .string(avatarUrl) }
        guard !updates.isEmpty else { return }

        try await supabase
            .from("profiles")
            .update(updates)
            .eq("id", value: userId)
            .execute()

        await fetchProfile()
    }
}
